pocld/shared_cl_context.cc

```diff
@@ -2117,16 +2121,39 @@ int SharedCLContext::createBuffer(BufferId_t BufferID, size_t Size,
   // handling is done in the client side), we cannot use any HOST_* flags here.
   Flags = Flags & (cl_bitfield)(CL_MEM_READ_WRITE | CL_MEM_WRITE_ONLY |
                                 CL_MEM_READ_ONLY);
-  // when RDMA is used, VirtualClContext passes in a pointer from clSVMAlloc
-  Flags = Flags |
-          (cl_bitfield)(HostPtr ? CL_MEM_USE_HOST_PTR : CL_MEM_ALLOC_HOST_PTR);
+
+  if (ParentID == 0) {
+    // When RDMA is used, VirtualClContext passes in a pointer from clSVMAlloc
+    Flags = Flags | (cl_bitfield)(HostPtr ? CL_MEM_USE_HOST_PTR
+                                          : CL_MEM_ALLOC_HOST_PTR);
+    // Note: Sub-buffer creation cannot have the HOST_PTR flags set.
+  }
 
   cl_int Err;
-  clBufferPtr Buf(
-      new cl::Buffer(ContextWithAllDevices, Flags, Size, HostPtr, &Err));
+  clBufferPtr Buf;
+
+  if (ParentID == 0) {
+    Buf = clBufferPtr(
+        new cl::Buffer(ContextWithAllDevices, Flags, Size, HostPtr, &Err));
+  } else {
+    std::unique_lock<std::mutex> Lock(BufferMapMutex);
+    assert(BufferIDmap.find(ParentID) != BufferIDmap.end());
+    clBufferPtr &Parent = BufferIDmap[ParentID];
```

Comment:
This unnecessarily re-does the lookup. Better to store the result from find in a variable before the assert instead and then do `clBufferPtr &Parent = FindResult->second` afterwards.

Also, since the BufferIDMap is not actually modified here I think it's safe to release the lock immediately after the find & assert?

Reply:
I've understood STL doesn't guarantee thread safety even for reads of the container in case there might be mutual updates. Better be safe than sorry with potential races and not micro-optimize prematurely.